mlabs/lendex-endpoint-spec.md

```diff
@@ -0,0 +1,205 @@
+
+# Lendex Endpoints Specification
+
+This document describes the endpoint-level bahaviours we want to guarantee for the Lendex Contract
+
+This will include every current/planned endpoint, and for each Endpoint, a list of:
+- input type
+- prerequiste actions/endpoint calls (implicit state required for the call to succeed on both wallet and script state, implied previous endpoint calls from the user)
+- user role qualifications, if any
+- expected effects
+- known invariant effects/invalid inputs
+- known risks which must be mitigated
+
+
+
+## About the Lendex Contract
+
+The Lendex Contract is a cardano adaptation of the Aave lending protocol.
+
+In Aave, tokens are used to provide the infrastructure of a savings & loan bank on the blockchain.
+
+Aave supports multiple currencies, and is mainly used to create Collateralized Debt Positions across currencies,   Aave will have individual Lendex contract instances available for each currency.
+
+A user can deposit the currency tied to a given Lendex Contract Instance, and mint a token called an `aToken` (so for `Ada`, `aAda`, `USD`, `aUSD` etc)
+
+These `aToken`s can then be supplied as Collateral against a loan (usually in a different currency)
+
+when interest is paid against the loan, this increases the supply of the underlying token, while no additional aTokens are minted, the result is that aToken value has increased, as the aToken exchange rate corresponds to the ratio of (all aTokens in circulation) : (all underlying tokens in the lendex contract).
+
+On user roles (Actor types):
+Almost all users will be suppliers within the system, being a supplier is _required_ to be a borrower, though many users who act as a supplier will never be borrowers.
+
+Additionally since LQ will be distributed among suppliers, Many (though not necessarily all) Governor users will also be suppliers.
+
+Liquidators need not be Suppliers though.
+
+Attackers will assume other roles, attempt to cause invariant behaviours, or take advantage of system rules such that they profit at the expense of the contract or it's users, we want to prevent these kinds of attacks or make them less profitable
+
+
+## Admin Contract
+
+### StartParams
+
+input: Mlabs.Lending.Contract.Api.StartParams
+
+prerequisite: none
+
+behaviour: instantiate a User Contract with the input as its initial parameters.
+
+### AddReserve
+
+input: Mlabs.Lending.Contract.Api.AddReserve
+
+Prerequisite: none
+
+behaviour: 
+
+-- @anton:  what does this endpoint do?
```

Comment:
It initializes reserve for given currencies.

Reply:
maybe my work on other lendex platforms has confused me, can you elaborate on where the reserve is used in the system? or is it just a safety of some kind?

Reply:
is the reserve in the aToken or the underlying?

Reply:
Reserve keeps all internal wallets and accounts. How much liquidity we have a lot of things.

see https://github.com/mlabs-haskell/plutus-use-cases/blob/main/mlabs/src/Mlabs/Lending/Logic/Types.hs#L91

Reply:
@anton-k - i'm afraid i'm still not quite sure why this is necessary, does a user also deposit funds to a reserve, which are used for loans? how is this different than regular deposits?

Reply:
It initializes the reserve. We can init the reserve and give funds to it. On the plutus level we keep all deposits attached to the same script. But we keep accounting which funds belong to which reserve in the reserve's datum.